apps/src/blockly/addons/cdoUtils.ts

```diff
@@ -318,11 +318,48 @@ export function isWorkspaceReadOnly() {
   return false; // TODO - used for feedback
 }
 
-export function blockLimitExceeded() {
-  return false;
+/**
+ * Checks if any block type's usage count exceeds its defined limit and returns
+ * the type of the first block found to exceed.
+ * @returns {string | undefined} The type of the first block that exceeds its limit,
+ * or undefined if no block exceeds the limit.
+ */
+export function blockLimitExceeded(): string | undefined {
+  const blockLimitMap = Blockly.blockLimitMap;
+  const blockCountMap = Blockly.blockCountMap;
+
+  // Ensure both maps are defined
+  if (!blockLimitMap || !blockCountMap) {
+    return undefined;
+  }
+
+  // Iterate over the block count map
+  for (const [type, count] of blockCountMap) {
+    const limit = blockLimitMap.get(type);
+    // If a limit is defined and the count exceeds this limit, return the type
+    if (limit !== undefined && count > limit) {
+      return type;
+    }
+  }
+
+  // If no count exceeds the limit, return undefined
+  return undefined;
 }
 
-export function getBlockLimit() {
+/**
+ * Retrieves the block limit for a given block type from the block limit map.
+ * @param {string} type The type of the block to check the limit for.
+ * @returns {number} The limit for the specified block type, or 0 if not found.
+ */
+export function getBlockLimit(type: string): number {
+  const blockLimitMap = Blockly.blockLimitMap;
+
+  // Check if the block limit map is defined and has the requested type
+  if (blockLimitMap && blockLimitMap.has(type)) {
+    return blockLimitMap.get(type) || 0;
+  }
+
+  // Return undefined if the map is not defined or the type is not found
```

Comment:
I think this comment is inaccurate since it references returning `undefined`. Also, what you have here might be the same as:

```
return Blockly.blockLimitMap?.get(type) ?? 0;
```

Reply:
Nice!